### PR TITLE
[expo-asset] Add native expo asset module with downloadAsync function.

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

To decouple expo-asset from expo-file-system and remove dependency.

If expo-assets needs a dependency on file-system, it makes it harder to maintain file-system and change the API of it, and we're planning to implement expo-file-system@next soon, aiming to possibly deprecate old functions sometime this year.

In addition, we're removing dependencies from the core `expo` package to make it lighter (see https://github.com/expo/expo/pull/27069 removing file-system as a dependency). In the future we might get rid of the expo-asset dependency as well. 

# How

Implemented native downloadAsync in swift + kotlin.

# Test Plan

Ensured unit tests for assets and file-system still pass on both platforms.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-asset/ios/AssetModule.podspec
Comment (tsapeta) [inline packages/expo-asset/ios/AssetModule.podspec:1]: Rename `AssetModule.podspec` to `ExpoAsset.podspec`

File: packages/expo-asset/ios/AssetModule.podspec [hunk 1]
```diff
@@ -0,0 +1,27 @@
+require 'json'
+
+package = JSON.parse(File.read(File.join(__dir__, '..', 'package.json')))
+
+Pod::Spec.new do |s|
+  s.name           = 'AssetModule'
```
Comment (tsapeta) [inline packages/expo-asset/ios/AssetModule.podspec:6]:
```suggestion
  s.name           = 'ExpoAsset'
```

File: packages/expo-asset/ios/AssetModule.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+internal class UnableToDownloadAssetException: GenericException<URL> {
+  override var reason: String {
+    "Unable to download asset from url: \(param)"
+  }
+}
+
+public class AssetModule: Module {
+  public func definition() -> ModuleDefinition {
+    Name("AssetModule")
```
Comment (tsapeta) [inline packages/expo-asset/ios/AssetModule.swift:12]:
```suggestion
    Name("ExpoAsset")
```

File: packages/expo-asset/package.json [hunk 3]
```diff
@@ -15,7 +15,9 @@
     "test": "expo-module test",
     "prepare": "expo-module prepare",
     "prepublishOnly": "expo-module prepublishOnly",
-    "expo-module": "expo-module"
+    "expo-module": "expo-module",
+    "open:ios": "open -a \"Xcode\" example/ios",
+    "open:android": "open -a \"Android Studio\" example/android"
```
Comment (tsapeta) [inline packages/expo-asset/package.json:20]: It doesn't have the example app
```suggestion
    "expo-module": "expo-module"
```

File: packages/expo-asset/package.json [hunk 4]
```diff
@@ -38,16 +40,15 @@
   },
   "dependencies": {
     "@react-native/assets-registry": "~0.73.1",
-    "blueimp-md5": "^2.10.0",
     "expo-constants": "~15.4.0",
     "expo-file-system": "~16.0.0",
     "invariant": "^2.2.4",
-    "md5-file": "^3.2.3"
+    "md5-file": "^3.2.3",
+    "expo-module-scripts": "^3.4.0",
```
Comment (tsapeta) [inline packages/expo-asset/package.json:47]: `expo-module-scripts` should always be a dev dependency

File: packages/expo-asset/package.json [hunk 5]
```diff
@@ -38,16 +40,15 @@
...
+    "expo-modules-core": "^1.11.2"
```
Comment (tsapeta) [inline packages/expo-asset/package.json:48]: Nothing should depend on `expo-modules-core` directly. Add a peer dep on `expo` instead

File: packages/expo-asset/src/AssetModule.web.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+export default {
+  downloadAsync: async function downloadAsync(
```
Comment (tsapeta) [inline packages/expo-asset/src/AssetModule.web.ts:2]:
```suggestion
  async downloadAsync(
```

**Comment by wschurman** [comment]
Would be helpful to detail a bit more about why we want to decouple these in the summary.

File: packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      }
+
+      val digest = md.digest()
+      fis.close()
+
+      val result = StringBuilder()
+      for (b in digest) {
+        result.append(String.format("%02x", b))
+      }
+      return result.toString()
```
Comment (lukmccall) [inline packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt:42]:
```suggestion
      return DigestInputStream(fis, md).use { dis ->
  	    while (dis.read() != -1) {
	        // Read the file content
	      }
	
	      md.digest().joinToString(separator = "") { "%02x".format(it) }
      }
```
It's safer to close streams with the `use` function. Also, if you close the `DigestInputStream` it should automatically close the `FileInputStream`, which was used during construction.

File: packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+import java.security.DigestInputStream
+import java.security.MessageDigest
+
+internal class UnableToDownloadAssetException(url: String) :
+  CodedException("Unable to download asset from url: $url")
+
+class AssetModule : Module() {
+  private fun getMD5HashOfFilePath(uri: URI): String {
+    val md = MessageDigest.getInstance("MD5")
+    return md.digest(uri.toString().toByteArray()).joinToString("") { "%02x".format(it) }
```
Comment (lukmccall) [inline packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt:22]:
```suggestion
    return md.digest(uri.toString().toByteArray()).joinToString(separator = "") { "%02x".format(it) }
```
I think adding the argument name makes that function much more readable.

File: packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    if (appContext.filePermission?.getPathPermissions(appContext.reactContext, localUrl.parent)?.contains(Permission.WRITE) != true) {
+      promise.reject(UnableToDownloadAssetException(uri.toString()))
+      return
+    }
+
+    try {
+      uri.toURL().openStream().use { input ->
+        localUrl.outputStream().use { output ->
+          input.copyTo(output)
+        }
```
Comment (lukmccall) [inline packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt:63]: Those operations are blocking, so can we dispatch them onto the background thread using coroutines? I think there should be a background coroutine scope available via the AppContext object.

**Comment by aleqsio** [comment]
@wschurman I've added some more context behind the PR – sorry for it being vague to start with. Let me know if you have any more questions!

File: packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  private fun getMD5HashOfFileContent(file: File): String? {
+    return try {
+      val md = MessageDigest.getInstance("MD5")
+      val fis = FileInputStream(file)
+      DigestInputStream(fis, md).use { dis ->
+        while (dis.read() != -1) {
+          // Read the file content
+        }
+        md.digest().joinToString(separator = "") { "%02x".format(it) }
+      }
```
Comment (wschurman) [inline packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt:35]: not sure doing the read of the file content manually is required. Here's what we use in expo-updates:

```suggestion
      FileInputStream(file).use { inputStream ->
        DigestInputStream(
          inputStream,
          MessageDigest.getInstance("MD5")
        ).use { digestInputStream ->
          digestInputStream.messageDigest.digest().joinToString(separator = "") { "%02x".format(it) }
        }
      }
```

File: packages/expo-asset/ios/AssetModule.swift [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+      return nil
+    }
+    return getMD5Hash(fromData: urlData)
+  }
+
+  private func getMD5Hash(fromData data: Data?) -> String? {
+    guard let data = data else {
+      return nil
+    }
+    return Data(Insecure.MD5.hash(data: data)).base64EncodedString()
```
Comment (wschurman) [inline packages/expo-asset/ios/AssetModule.swift:54]: at a glance it seems that android is using hex encoding while iOS is using base64?

**Comment by aleqsio** [comment]
@tsapeta The instrumentation tests and maestro both seem to be having issues, but both seem unrelated, right?

File: packages/expo-asset/ios/AssetModule.swift [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    do {
+      try appContext.fileSystem?.ensureDirExists(withPath: localUrl.path)
+    } catch {
+      promise.reject(UnableToDownloadAssetException(url))
+      return
+    }
+    guard let fileSystem = appContext.fileSystem else {
+      promise.reject(UnableToDownloadAssetException(url))
+      return
+    }
```
Comment (tsapeta) [inline packages/expo-asset/ios/AssetModule.swift:63]: I would move this up, before `do..catch` as you already use `fileSystem` there

File: packages/expo-asset/ios/AssetModule.swift [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+      return nil
+    }
+    return Data(Insecure.MD5.hash(data: data)).map { String(format: "%02hhx", $0) }.joined()
+  }
+
+  func downloadAsset(appContext: AppContext, url: URL, localUrl: URL, promise: Promise) {
+    do {
+      try appContext.fileSystem?.ensureDirExists(withPath: localUrl.path)
+    } catch {
+      promise.reject(UnableToDownloadAssetException(url))
```
Comment (tsapeta) [inline packages/expo-asset/ios/AssetModule.swift:57]: Would be nice to use a bit more specific exception. If we get here, it's not the problem with downloading itself but ensuring the directory exists.

File: packages/expo-asset/ios/AssetModule.swift [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+      return
+    }
+
+    let downloadTask = URLSession.shared.downloadTask(with: url) { urlOrNil, _, _ in
+      guard let fileURL = urlOrNil else {
+        promise.reject(UnableToDownloadAssetException(url))
+        return
+      }
+      do {
+        try? FileManager.default.removeItem(at: localUrl)
```
Comment (tsapeta) [inline packages/expo-asset/ios/AssetModule.swift:75]: If the optional try is intentional, can you add a comment why?

File: packages/expo-modules-core/CHANGELOG.md [hunk 15]
```diff
@@ -27,6 +27,7 @@
 
 ### 💡 Others
 
+- Deprecated `expo.modules.core.Promise`. ([#27369](https://github.com/expo/expo/pull/27369) by [@aleqsio](https://github.com/aleqsio))
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:30]: These changes should be made in a separate PR 😉 You're linking to a PR that has nothing to do with the deprecation.

**Review by tsapeta — CHANGES_REQUESTED** [review]

File: packages/expo-asset/ios/AssetModule.swift (hunk 14 above)
Comment (aleqsio) [reply packages/expo-asset/ios/AssetModule.swift:82]: Added a comment 
```
        // the file may already exist, so we need to remove it first
```

File: apps/expo-go/ios/Podfile.lock [hunk 16]
```diff
@@ -2021,7 +2027,7 @@ SPEC CHECKSUMS:
   GoogleDataTransport: 54dee9d48d14580407f8f5fbf2f496e92437a2f2
   GoogleMaps: a146f275ee429d14822178c7a841c03366ec92a1
   GoogleUtilities: 13e2c67ede716b8741c7989e26893d151b2b2084
-  hermes-engine: 0933c22606de894dccc3248ae93129951a6e8e6b
```
Comment (aleqsio) [inline apps/expo-go/ios/Podfile.lock:2024]: it bumped hermes for me after removing Pods – should I do something else to avoid it?

Comment (tsapeta) [reply apps/expo-go/ios/Podfile.lock:2024]: Ok, let's ignore it then. Maybe it was just not correct in Expo Go 🤷

**Review by tsapeta — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]
